File: javascript/src/problems301_400/intersection-of-two-arrays2.js
class Problem {
  intersect(nums1, nums2) {
    throw new Error(`Not implemented (${nums1}, ${nums2}).`);
  }
}

class Default extends Problem {
  intersect(nums1, nums2) {
    nums1.sort((a, b) => a - b);
    nums2.sort((a, b) => a - b);

    let i = 0;
    let j = 0;
    const result = [];
    while (i < nums1.length && j < nums2.length) {
      if (nums1[i] < nums2[j]) {
        i++;
      } else if (nums1[i] > nums2[j]) {
        j++;
      } else {
        result.push(nums1[i]);
        i++;
        j++;
      }
    }
    return result;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems301_400/intersection_of_two_arrays2.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def intersect(self, nums1: list[int], nums2: list[int]) -> list[int]:
        pass


class Default(Problem):
    @override
    def intersect(self, nums1: list[int], nums2: list[int]) -> list[int]:
        nums1.sort()
        nums2.sort()

        i: int = 0
        j: int = 0
        result: list[int] = []
        while i < len(nums1) and j < len(nums2):
            if nums1[i] < nums2[j]:
                i += 1
            elif nums1[i] > nums2[j]:
                j += 1
            else:
                result.append(nums1[i])
                i += 1
                j += 1
        return result


SOLUTIONS: list[Problem] = [Default()]
